Use 'open_in_bin', 'open_out_bin' instead of 'open_in', 'open_out'

mjambon opened the pull request:
I replaced `open_in` by `open_in_bin` everywhere so as to avoid contents mangling on Windows. This will avoid crashing in some cases (e.g. when reading marshaled data) and will avoid invalid locations or `End_of_file` errors in other cases. Parsers are already capable of interpreting both CRLF and LF as line terminators anyway.

This is most of the work for https://github.com/returntocorp/semgrep/issues/3061
Might fix (?) https://github.com/returntocorp/semgrep/issues/2991

mjambon committed:
Use 'open_in_bin' instead of 'open_in' everywhere so as to avoid
contents mangling on Windows. This will avoid crashing in some cases
(e.g. when reading marshaled data) and will avoid invalid locations or
End_of_file errors in other cases. Parsers are already capable of
interpreting both CRLF and LF as line terminators anyway.

mjambon commented on commons/Common.ml:
```diff
@@ -446,7 +446,7 @@ let profile_code2 category f =
-  let chan = open_in filename in
+  let chan = open_in_bin filename in
   let x = input_value chan in (* <=> Marshal.from_channel  *)
   (close_in chan; x)
```
This would have failed (segfault) on Windows.

aryx replied:
Nice

mjambon commented on commons/Common.ml:
```diff
@@ -922,7 +922,7 @@ let cat file =
+  let ic = open_in_bin file  in
   let size = in_channel_length ic in
   let buf = Bytes.create size in
   really_input ic buf 0 size;
```
I suspect this could attempt to read beyond the end of the file, presumably resulting in an `End_of_file` exception on Windows.

This is because `in_channel_length` does this:
> The returned size does not take into account the end-of-line translations that can be performed when reading from a channel opened in text mode.

and I assume `really_input` does perform the CRLF->LF conversions because the documentation doesn't say otherwise.

Frankly I don't know which conversions happen and when because (1) the documentation isn't clear about it and (2) I don't have a Windows machine. That's why it's easier to open all files in binary mode.

Source: https://ocaml.org/api/Stdlib.html

aryx replied:
maybe you should add this comment in the code, so we remember, because I also don't understand anything
about those binary vs textual issue in Windows.

mjambon replied:
What Windows does isn't complicated, it's the OCaml implementation on top of the Windows syscalls that's not obvious.

I'll explain this in `Common.ml` or maybe in a separate readme. What the Windows syscalls do on files opened in text mode is:
* reads: any CRLF (`\r\n`) sequence is translated into LF (`\n`).
* writes: any LF is translated into CRLF.

From that follows a number of rules on how to write our application, I'll detail those as well.

mjambon replied:
I'll also explain the situation with stdin, stdout, and stderr (the open files, not the file descriptors), which are always open in text mode.

mjambon replied:
I changed the behavior of `cat` and added a big explanation in the middle of `Common.ml` because I couldn't think of a better place.

mjambon commented:
The story is similar for `open_out` and `open_out_bin`. Working on it.

mjambon committed:
Open all files for writing in binary mode so as to get the same output
on Unix and Windows. The stdout and stderr are already opened like
they should on the local platform, so we don't need to worry about them
when writing to the console.

mjambon commented:
I switched to `open_out_bin` as well. There may be situations where Windows users want to see a log file with CRLF line endings rather than LF. We can fine-tune that when the time comes.

aryx commented on commons/Common.ml:
```diff
@@ -911,7 +911,7 @@ let cmd_to_list_and_status = process_output_to_list2
 
 (* tail recursive efficient version *)
 let cat file =
-  let chan = open_in file in
```
I understand the use of open_in_bin to read non-textual data, but then for 'cat' here we know it's textual data,
so why do we need open_in_bin also?

mjambon replied:
`Stdlib.input_line` behaves ~~incorrectly~~ not like we want on Windows files read on Unix or in binary mode. I fixed that and added a test.

mjambon committed:
Modify the 'cat' function to exclude CR from the line when reading a
Windows file on Unix, like parsers ought to do.
Add a general explanation of CRLF vs. LF issues.

mjambon committed:
Add test for the 'cat' implementation that uses the CRLF-compatible
implementation of 'input_line'.
